sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala

```diff
@@ -49,8 +48,6 @@ abstract class RDG extends UnaryExpression with ExpectsInputTypes with Stateful
   @transient protected lazy val seed: Long = seedExpression match {
     case Literal(s, IntegerType) => s.asInstanceOf[Int]
     case Literal(s, LongType) => s.asInstanceOf[Long]
```

Comment:
Ur, we might need to update the two entries above like this;
```
    case e if e.dataType == IntegerType => e.eval().asInstanceOf[Int]
    case e if e.dataType == LongType => e.eval().asInstanceOf[Long]
```
Otherwise, we can hit a unmatch exception when excluding `ConstantFolding` and `ReorderAssociativeOperator`?
```
// v3.0.1
:$./bin/spark-shell --conf spark.sql.optimizer.excludedRules=org.apache.spark.sql.catalyst.optimizer.ConstantFolding,org.apache.spark.sql.catalyst.optimizer.ReorderAssociativeOperator
scala> sql("select rand(1 + 2)").show()
org.apache.spark.sql.AnalysisException: Input argument to rand must be an integer, long or null literal.;
  at org.apache.spark.sql.catalyst.expressions.RDG.seed$lzycompute(randomExpressions.scala:49)
  at org.apache.spark.sql.catalyst.expressions.RDG.seed(randomExpressions.scala:46)
  at org.apache.spark.sql.catalyst.expressions.Rand.doGenCode(randomExpressions.scala:98)
  at org.apache.spark.sql.catalyst.expressions.Expression.$anonfun$genCode$3(Expression.scala:146)
  at scala.Option.getOrElse(Option.scala:189)
```

Reply:
thanks @maropu , I have checked and updated it.

Reply:
I thin it is bad that the argument of `rand` depends on the optimizer rules and it is a minor bug, so could you fix it in a separate PR first for backporting the fix? cc: @HyukjinKwon @dongjoon-hyun

Reply:
It might a little diffcult to fix this in a separate PR. We also need check seed is foldable first if we use `eval` instead of `match`, otherwise the seed can be something like `udf(): Int` which is not unpected.

Reply:
also cc @cloud-fan